src/main/java/frc/robot/utils/MovingAverage.java

```diff
@@ -0,0 +1,114 @@
+
+    public MovingAverage(String pathToCsv) {
+        this(new File(pathToCsv));
+    }
+
+    public MovingAverage(File csv) {
+        this.distanceVelocityMap = new TreeMap<>();
+        if (csv.exists() && csv.isFile()) {
+            try {
+                read(new FileReader(csv));
+            } catch (FileNotFoundException e) { // Can't happen, we have already checked it
+                e.printStackTrace();
+            }
+        } else
+            System.err.println("File not found: " + csv.getName());
+    }
+
+    public MovingAverage(Map<Double, Double> map) {
+        this.distanceVelocityMap = new TreeMap<>(map); // just because I want to order the values by the key.
+    }
+
+    public MovingAverage(double[][] values) {
+        this.distanceVelocityMap = new TreeMap<>();
+        for (double[] pair : values) {
+            if (pair.length >= 2)
+                this.distanceVelocityMap.put(pair[0], pair[1]);
+        }
+    }
+
+    public MovingAverage(Reader reader) {
+        this.distanceVelocityMap = new TreeMap<>();
+        read(reader);
+    }
+
+    private void read(Reader reader) {
+        String line = null;
+        try (BufferedReader bufferedReader = new BufferedReader(reader)) {
+            while ((line = bufferedReader.readLine()) != null) {
+                String[] pair = line.split(DELIMITER);
+                this.distanceVelocityMap.put(Double.parseDouble(pair[0]), Double.parseDouble(pair[1]));
+            }
+        } catch (IOException e) {
+            System.err.println("MovingAverage class can't read the file");
+            System.out.println(e.getMessage());
+        } catch (IndexOutOfBoundsException e) {
+            System.err.println("Please check that every line has a pair of values, the problem occurred on this line " + line);
+        }
+    }
+
+    /**
+     * Retrieve an approximation for the velocity based on the distance.
+     *
+     * @param distance the distance from the target.
+     * @return an approximation of the velocity to apply
+     */
+    public double estimateVelocity(double distance) {
+        double[][] nearest = getClosestDistance(distance);
+
+        if (nearest[0][0] == nearest[1][0]) // we have already have a record of the distance, just return the velocity based of that
+            return nearest[0][1];
+
+        // y = mx + b
+        // m = delta y / delta x
+        double m = (nearest[1][1] - nearest[0][1]) / (nearest[1][0] - nearest[0][0]);
+        // b = y - m*x
+        double b = nearest[1][1] - m * nearest[1][0];
+
+        return distance * m + b;
+    }
+
+    /**
+     * get a 2 pairs of (distance, velocity) of the closets distances below and above.
+     *
+     * @param distance the distance to return the pairs.
+     * @return a 2d array represents the closest values(below {distance, velocity}, above {distance, velocity}).
+     */
+    private double[][] getClosestDistance(double distance) {
```

Comment:
ask @orel-david why we use this approach

Reply:
And about the first part?

Reply:
I still can do it using the map, I just need to use the [keySet](https://docs.oracle.com/en/java/javase/15/docs/api/java.base/java/util/Map.html#keySet()) function, and I can treat it the same way

Reply:
Could you elaborate? I recommended using an array so if you find the distance before the given distance, it's easy to find the next one because you can access values by index.

Reply:
```java
Set<Double> keys = distanceVelocityMap.keySet();
for (var iterator = keys.iterator(); iterator.hasNext();) {
    double val = iterator.next(); // automatically change the index...
    if (<logic>)
         return new double[] {val, iterator.next()};
}
```

Reply:
Oh, very nice 🙂. I didn't know you could do that.